### PR TITLE
gh-101476: Add _PyType_GetModuleState

### LINKED ISSUE #101476: Add internal API for fast module access from heap type methods
See [topic on Discourse](https://discuss.python.org/t/a-fast-variant-of-pytype-getmodule/23377?u=erlendaasland).

For CPython internal usage, we've got `_PyModule_GetState`, which is a fast variant of `PyModule_GetState`, the module check in the latter is simply an assert in the former.

For `PyType_GetModuleState`, there are three `if`s (two of them implicitly in `PyType_GetModule`):

1. check that the given type is a heap type
2. check that the given type has an associated module
3. check that the result of `PyType_GetModule` is not `NULL`

For stdlib core extension modules, all of these conditions are always true (AFAIK). With a fast static inlined variant, for example `_PyType_GetModuleState`, with a fast variant of `PyType_GetModule` inlined, where all three conditions are `assert()`ed, we can speed up a heap type methods that need to access module state.

### Linked PRs
* gh-101477
* gh-102188

### PR DESCRIPTION
**Pull request by erlend-aasland**
For fast module state access from heap type methods.

* Issue: gh-101476

### CODE REVIEW

**Comment by erlend-aasland** [comment]
I'm not sure we should add a NEWS entry for this; this is an internal C API.

**Review by kumaraditya303 — APPROVED** [review]
Nice!

File: Include/internal/pycore_typeobject.h
Comment (kumaraditya303) [inline Include/internal/pycore_typeobject.h]:
```suggestion
    PyHeapTypeObject *et = (PyHeapTypeObject *)type;
    assert(et->ht_module);
    PyModuleObject *mod = (PyModuleObject *)(et->ht_module);
```

Comment (erlend-aasland) [reply Include/internal/pycore_typeobject.h]: Thanks!

**Comment by erlend-aasland** [comment]
Thanks for the review, Kumar!

Even though I also got thumbs-up from Guido og Steve on Discourse, I'd like to wait for at least one other core dev approval before landing this PR.

**Comment by encukou** [comment]
+1 from me, could you just add a comment to clarify the intent, e.g. `/* Like PyType_GetModuleState, but skips verification that type is a heap type with an associated module */`

**Comment by erlend-aasland** [comment]
> +1 from me, could you just add a comment to clarify the intent, e.g. `/* Like PyType_GetModuleState, but skips verification that type is a heap type with an associated module */`

Thanks for the review, Petr. I'll add the comment right away.

**Review by encukou — APPROVED** [review]

**Review by kumaraditya303 — APPROVED** [review]

File: Include/internal/pycore_typeobject.h
Comment (gpshead) [inline Include/internal/pycore_typeobject.h:70]: I trust your need for the API based on the discuss thread. The only question I have is if this should use the ALL_CAPS idiom to represent the "unchecked fast API" as other things such as `_PyObject_GET_WEAKREFS_LISTPTR` and `PyTuple_GET_ITEM` do.

`_PyType_GET_MODULE_STATE()` perhaps?

That makes the difference from PyType_GetModuleState more obvious than just the leading underscore.

Comment (erlend-aasland) [reply Include/internal/pycore_typeobject.h:70]: Yeah, that makes sense to me.

Ideally, `_PyModule_GetState` should have followed the same convention (`_PyModule_GET_STATE`).

**Review by gpshead — APPROVED** [review]

File: Include/internal/pycore_typeobject.h
Comment (encukou) [reply Include/internal/pycore_typeobject.h:70]: That's the naming convention for macros (see [PEP 7](https://peps.python.org/pep-0007/#naming-conventions)), and -- I suppose -- things that started out as macros but were converted to functions.
Using it to mean “unchecked fast API” surprised me a bit. Should we update PEP 7?

Comment (erlend-aasland) [reply Include/internal/pycore_typeobject.h:70]: Post PEP-670 it may be a good idea to update PEP-7, yeah. Pre PEP-670 this would probably have been a macro, I guess, so the naming kinda makes sense. I'm fine with both `_PyType_GetModuleState` and `_PyType_GET_MODULE_STATE`.

Comment (erlend-aasland) [reply Include/internal/pycore_typeobject.h:70]: Revisiting this; I think we should go for the original proposal, `_PyType_GetModuleState`, which means reverting c76260ae07213a2cceee67a0622f167b80d5d091

Comment (encukou) [reply Include/internal/pycore_typeobject.h:70]: FWIW, I don't really care either way -- it's private API. I was just surprised when it was brought up.

**Comment by erlend-aasland** [comment]
Thanks for the reviews, everyone! I'm applying this across the core extension modules. See #102188.

File: Include/internal/pycore_typeobject.h
Comment (vstinner) [inline Include/internal/pycore_typeobject.h:77]: This assertion is the same than ``assert(et->ht_module);``, no? If you only keep one, I suggest keeping ``assert(mod != NULL);`` since we use mod below.

File: Include/internal/pycore_typeobject.h
Comment (vstinner) [reply Include/internal/pycore_typeobject.h:70]: If a function is exposed to the public C API, for "unsafe" flavor of a function, I would suggest adding "Unsafe" or "Unchecked" to its name rather than relying on lowercase/uppercase. Example of existing functions:

* \_PyThreadState_**Unchecked**Get()
* \_PyGILState_GetInterpreterState**Unsafe**()

For the specific case of the internal C API, so far, I just added ``_`` prefix: ``PyThreadState_GET()`` (slow, regular function call) becomes  ``_PyThreadState_GET()`` (fast, inlined). I'm not really excited by a single underscore, it's easy to miss. But well, that's the status quo convention :-)

Maybe we should just go for ``Fast`` suffix :-D ``_PyType_GetModuleStateFast()`` :-)

File: Include/internal/pycore_typeobject.h
Comment (erlend-aasland) [reply Include/internal/pycore_typeobject.h:77]: ... or the first assert could be changed to `assert(et != NULL)`.

**Pull request merged**
